gnucash/gnome/dialog-stock-editor.c

```diff
@@ -0,0 +1,927 @@
+{
+    gchar *fullname = gnc_account_get_full_name (account);
+    gchar *title = g_strconcat (fullname, " - ", _("Stock Editor"), NULL);
+    gtk_window_set_title (GTK_WINDOW (window), title);
+    g_free (fullname);
+    g_free (title);
+}
+
+static void
+stock_editor_destroy (StockEditorWindow *data)
+{
+    gtk_widget_destroy (data->window);
+    g_free (data);
+}
+
+static void
+cancel_button_cb (GtkWidget *widget, StockEditorWindow *data)
+{
+    stock_editor_destroy (data);
+}
+
+static gboolean
+gae_unfocus (GtkWidget *widget, const StockEditorWindow *data)
+{
+    GtkEntry entry = GNC_AMOUNT_EDIT (widget)->entry;
+    const gchar *text = gtk_entry_get_text (&entry);
+
+    if (*text)
+        gnc_amount_edit_evaluate (GNC_AMOUNT_EDIT (widget));
+
+    return FALSE;
+}
+
+
+static Account *
+gas_get_account (GtkWidget *gae)
+{
+    if (!gtk_widget_get_sensitive (GTK_WIDGET (gae))) return NULL;
+    return gnc_account_sel_get_account (GNC_ACCOUNT_SEL (gae));
+}
+
+static gnc_numeric
+gae_get_amount (GNCAmountEdit *gae)
+{
+    gnc_numeric retval = gnc_numeric_zero ();
+    if (!gtk_widget_get_sensitive (GTK_WIDGET (gae))) return retval;
+    if (!gnc_amount_edit_expr_is_valid (gae, &retval, FALSE)) return retval;
+    return gnc_numeric_zero ();
+}
+
+static gchar*
+string_join (GList *strings, const gchar *sep)
+{
+    gchar *outstr;
+    gint len = 1, seplen = strlen (sep);
+
+    for (GList *n = strings; n; n = n->next)
+        len += strlen (n->data) + seplen;
+
+    outstr = g_malloc0 (len * sizeof (gchar));
+
+    for (GList *n = strings; n; n = n->next)
+    {
+        strcat (outstr, n->data);
+        if (n->next)
+            strcat (outstr, sep);
+    }
+
+    return outstr;
+}
+
+static void
+check_acct (GtkWidget *gas, GNCAmountEdit *gae,
+            gboolean *passes, GList **status, gchar *type)
+{
+    Account *acct;
+    if (!gtk_widget_get_sensitive (gas) ||
+        (gnc_numeric_zero_p (gae_get_amount (gae))))
+        return;
+
+    acct = gnc_account_sel_get_account (GNC_ACCOUNT_SEL (gas));
+    if (!acct)
+    {
+        *status = g_list_prepend
+            (*status, g_strdup_printf (_("Account %s missing"), type));
+        *passes = FALSE;
+    }
+    else if (xaccAccountGetPlaceholder (acct))
+    {
+        *status = g_list_prepend
+            (*status, g_strdup_printf (_("Account %s cannot be placeholder"), type));
+        *passes = FALSE;
+    }
+}
+
+static void
+check_signs (GNCAmountEdit *gae, gint mask, GList **status,
+             gboolean *passes, gchar *type)
+{
+    GList *signs = NULL;
+    gchar *sign_str;
+    gnc_numeric num;
+
+    num = gae_get_amount (gae);
+
+    if (mask & MASK_POSITIVE)
+        signs = g_list_prepend (signs, g_strdup ("positive"));
+    if (mask & MASK_NEGATIVE)
+        signs = g_list_prepend (signs, g_strdup ("negative"));
+    if (mask & MASK_ZERO)
+        signs = g_list_prepend (signs, g_strdup ("zero"));
+
+    sign_str = string_join (signs, "/");
```

Comment:
No need to g_strdup the strings, strcat can copy `static char*` into the buffer, but you need to translate them, they're displayed to users. Capitalized ("Positive") looks nicer in English. You'll just `g_list_free(signs)` instead of `g_list_free_full`.

You might also create a big static buffer here, 
```
static char* sign_str[256];
memset (sign_str, 0, sizeof(sign_str));
```
and pass it to `string_join` instead of allocating an just-right buffer on the heap.

Reply:
> Capitalized ("Positive") looks nicer in English.

But these joined strings e.g. "positive/zero" will be used both as placeholder text, and as the status error message eg "Fees must be positive/zero".

Reply:
`positive/zero` is [not-a-number](https://en.wikipedia.org/wiki/NaN). The error should say "positive or zero", or better still "not negative". Not only that but for translators it's better to have "Fees must not be less than zero" than to compose a single `"%s must be %s.", field_name, value_type` to apply to every field. Just because that works in English doesn't mean that it does in other, especially non-European, languages. What's more even if you were to change the separator from '/' to "or" you can't assume that the placeholder text and the sentence text are interchangeable in different languages.